test/known_problems/should_pass/list_nth.erl

```diff
@@ -0,0 +1,32 @@
+-module(list_tail).
+
+-compile([export_all]).
+
+% Pass
+-spec nth1([integer()]) -> integer().
+nth1([]) -> 0;
+nth1(Xs) -> lists:nth(1, Xs).
+
+% Fail but should pass
+-spec nth2([integer()]) -> integer().
+nth2(Xs) ->
+    case length(Xs) > 0 of
+        true -> lists:nth(1, Xs);
+        false -> 0
+    end.
```

Comment:
I agree that there always will be programs which exceed the capabilities of Gradualizer. However if Gradualizer emits false errors in this case, that renders the tool unusable, because the users have to scan through noise in the output to find real errors.
Maybe one solution is if Gradualizer is aware of its limitations and in such cases falls back to the gradual type (or a subset of that), basically fails open. I think not finding type errors is much better user experience. And it can be still improved gradually so there are less and less such cases.

Reply:
If we do that, we get an unsound typesystem. Then, our use case is basically the same as for Dialyzer.

All programs can be divided into 3 categories:

1. Programs known to be free from type errors, "white list"
2. Programs known to contain type errors, "black list"
3. Programs for which it's unknown if they have type errors or not, "gray list"

The nice property of a *sound* type system is that if the type checker doesn't complain, there is a guarantee that it's free from type errors. This means that there will be no case_clause, function_clause, badarg, etc. errors in runtime.

(Now, for a gradual typesystem, this guarantee applies only to the parts of programs where the types are specified, but for the example here, everything is typed.)

Dialyzer never complains about programs in the gray list. That's nice in some ways, but on the other hand, you never have any guarantees. Gradualizer aims to be sound, so we *must* report type errors for programs on the gray list. Then, the programmer has to change the program into something that the typesystem can verify, e.g. nth1 or nth4. This is one of the main differences between Dialyzer and Gradualizer IMO.

(It is possible to narrow down the gray list as much as we can, for example by giving `length/1` the type `([]) -> 0; (nonempty_list()) -> pos_integer()` and then do something smart with `case length(Xs) > 0 of ...`, but there will still always be a gray list.)

Reply:
Agree. I want to add that it should always be possible to escape type errors without changing the behavior of the code. By having that, it is quite fine to be very strict and to not handle all programs. I think that the assertion system we have is already good enough to go this route.

Reply:
true, you guys are right